ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/util/ForkChoiceUtil.java

```diff
@@ -439,4 +440,33 @@ private boolean hasAncestorAtSlot(
         .map(ancestorAtSlot -> ancestorAtSlot.equals(ancestorRoot))
         .orElse(false);
   }
+
+  public boolean canOptimisticallyImport(final ReadOnlyStore store, final SignedBeaconBlock block) {
+    // A block can be optimistically imported either if it's parent contains a non-default payload
+    // or if it is at least `SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY` before the current slot.
+    // This is to avoid the merge block referencing a non-existent eth1 parent block and causing
+    // all nodes to switch to optimistic sync mode.
+    if (isExecutionBlock(store, block.getParentRoot())) {
+      return true;
+    }
+    return isBellatrixBlockOld(store, block.getSlot());
+  }
+
+  private boolean isBellatrixBlockOld(final ReadOnlyStore store, final UInt64 blockSlot) {
+    final Optional<SpecConfigBellatrix> maybeConfig = specConfig.toVersionBellatrix();
+    if (maybeConfig.isEmpty()) {
+      return false;
+    }
+    return blockSlot
+        .plus(maybeConfig.get().getSafeSlotsToImportOptimistically())
+        .isLessThanOrEqualTo(getCurrentSlot(store));
+  }
+
+  private boolean isExecutionBlock(final ReadOnlyStore store, final Bytes32 blockRoot) {
+    final Optional<Bytes32> parentExecutionRoot =
+        store.getForkChoiceStrategy().executionBlockHash(blockRoot);
+    final boolean isExecutionBlock =
+        parentExecutionRoot.isPresent() && !parentExecutionRoot.get().isZero();
+    return isExecutionBlock;
```

Comment:
nit: the local boolean `isExecutionBlock` seems a bit redundant to me. The method itself has that name and should be enough to understand the objective.

Reply:
IntelliJ agrees. :)